# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1786832936 -7200
# Parent  c04f8c2c22e21df9c47d0ecdfb1548278c9538c8
NNN7 - Fix wrong folder selection at startup.

Registed upstream as https://bugzilla.mozilla.org/show_bug.cgi?id=1875655.
Also fixes https://bugzilla.mozilla.org/show_bug.cgi?id=2063838.

When restoring the first tab, sometimes the top-most folder gets selected before
later the right folder is selected. Waiting before the "select" event dispatch
seems to fix this issue.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -110,8 +110,11 @@ var multiMessageBrowser;
  * is selected.
  */
 var accountCentralBrowser;
 
+var initialDisplayComplete = false;
+var suppress_gViewWrapper_close = false;
+
 /**
  * HTML body element handling the general layout of the about3pane.
  */
 var paneLayout;
@@ -212,12 +215,47 @@ window.addEventListener("DOMContentLoade
   // (triggered by `folderPane.init` and possibly `restoreState`) are ignored
   // to avoid unnecessarily loading the thread tree or Account Central.
   folderTree.addEventListener("select", folderPane);
 
-  // Delay initial folder selection until after the message list's resize
-  // observer has had a chance to respond to layout changes. Otherwise we
-  // might end up scrolling to the wrong part of the list.
-  await new Promise(resolve => setTimeout(resolve));
+  // In some cases `folderTree.selectedIndex` isn't set correctly when the event is
+  // handled, so the wrong folder is selected at first. We've seen
+  // index 0, so in a unified view, that will be the unified inbox as the top-most row.
+  // If that's grouped-by-sort, then long delays are expected.
+  // Fixing the long delays in https://bugzilla.mozilla.org/show_bug.cgi?id=1875577
+  // will paper over the issue some more.
+  // Sadly folderPane.init() initializes the index to 0, so if it's 0 here after
+  // `restoreState()`, either 0 was indeed the desired the selection or it's not
+  // correctly initialized yet. `restoreState()` looks like a sync function, but
+  // `initialDisplayComplete` isn't always `true` here. Note that `displayFolder()`
+  // sets the index to what is required.
+  let count = 0;
+  let timeWaited = 0;
+  while (count++ < 5 && !initialDisplayComplete) {
+    timeWaited += 100;
+    await new Promise(resolve => setTimeout(resolve, 100));
+  }
+  if (timeWaited) {
+    if (initialDisplayComplete) {
+      // eslint-disable-next-line no-console
+      console.info(
+        `3pane init: Waited ${timeWaited} ms and index of selected folder is ${folderTree.selectedIndex}`
+      );
+    } else {
+      console.warn(
+        `3pane init: Waited ${timeWaited} ms and index of selected folder is ${folderTree.selectedIndex}`
+      );
+    }
+  }
+
+  // Since Supernova kaboom, startup is a total mess of pingpong balls and mousetraps.
+  // Way before we get here, messages are fetched in messenger.js with a call to
+  // `MsgGetMessagesForAllServers()` in `loadStartFolder()`.
+  // The handler for the event we're about to dispatch calls
+  // `gViewWrapper?.close()` and that resets calls `clearNewMessages()`.
+  // Not nice when it runs on the folder we've already received message into.
+  // Debugging showed that in 140, `gViewWrapper` was `undefined`, so `close()` wasn't
+  // called on it. Here we enforce that state.
+  suppress_gViewWrapper_close = true;
   folderTree.dispatchEvent(new CustomEvent("select"));
 
   CalMetronome.on("day", refreshGroupedBySortView);
 
@@ -2961,8 +2999,11 @@ var folderPane = {
     }
   },
 
   async _onSelect() {
+    const suppress = suppress_gViewWrapper_close;
+    suppress_gViewWrapper_close = false;
+
     const isSynthetic = gViewWrapper?.isSynthetic;
     threadPane.saveSelection();
     threadPane.hideIgnoredMessageNotification();
     if (!isSynthetic) {
@@ -3014,9 +3055,11 @@ var folderPane = {
     document.head.querySelector(`link[rel="icon"]`).href =
       FolderUtils.getFolderIcon(gFolder);
 
     // Clean up any existing view wrapper. This will invalidate the thread tree.
-    gViewWrapper?.close();
+    if (!suppress) {
+      gViewWrapper?.close();
+    }
 
     if (gFolder.isServer) {
       document.title = gFolder.server.prettyName;
       gViewWrapper = gDBView = threadTree.view = null;
@@ -7440,9 +7483,11 @@ function restoreState({
   paneLayout.messagePaneSplitter.isCollapsed = !messagePaneVisible;
 
   if (folderURI) {
     displayFolder(folderURI);
+    initialDisplayComplete = true;
   } else if (syntheticView) {
+    initialDisplayComplete = true;
     // In a synthetic view check if we have a previously edited column layout to
     // restore.
     if ("getPersistedSetting" in syntheticView) {
       const columnsState = syntheticView.getPersistedSetting("columns");
